In [58]:
url= "https://raw.githubusercontent.com/DeepentiA/Capstone_Project/master/best%20food.csv"
df5 = pd.read_csv(url)
df5.head()

In [59]:
import requests
url ="https://raw.githubusercontent.com/DeepentiA/Capstone_Project/master/Gurgaon_dataset.csv"
df = pd.read_csv(url)
print(df)
df.head(20)

    Unnamed: 0          Borough         Neighborhoods   Latitude  Longitude  \
0            0          Central       Cantonment area  12.972442  77.580643   
1            1          Central                Domlur  12.960992  77.638726   
2            2          Central           Indiranagar  12.971891  77.641151   
3            3          Central     Jeevanbheemanagar  12.962900  77.659500   
4            4          Central           Malleswaram  13.003100  77.564300   
5            5          Central             Pete area  12.962700  77.575800   
6            6          Central           Rajajinagar  12.990100  77.552500   
7            7          Central        Sadashivanagar  13.006800  77.581300   
8            8          Central         Seshadripuram  12.993500  77.578700   
9            9          Central          Shivajinagar  12.985700  77.605700   
10          10          Central                Ulsoor  12.981700  77.628600   
11          11          Central         Vasanth Naga

Unnamed: 0  Borough      Neighborhoods   Latitude  Longitude  Population  \
0            0  Central    Cantonment area  12.972442  77.580643      866377   
1            1  Central             Domlur  12.960992  77.638726      743186   
2            2  Central        Indiranagar  12.971891  77.641151      474289   
3            3  Central  Jeevanbheemanagar  12.962900  77.659500      527874   
4            4  Central        Malleswaram  13.003100  77.564300      893629   
5            5  Central          Pete area  12.962700  77.575800      730999   
6            6  Central        Rajajinagar  12.990100  77.552500      981362   
7            7  Central     Sadashivanagar  13.006800  77.581300      662625   
8            8  Central      Seshadripuram  12.993500  77.578700      396862   
9            9  Central       Shivajinagar  12.985700  77.605700       77836   
10          10  Central             Ulsoor  12.981700  77.628600      656726   
11          11  Central      Vasanth Nagar  12.991100  77.592000      942711   
12          12  Eastern          Bellandur  12.926000  77.676200      208094   
13          13  Eastern     CV Raman Nagar  12.985500  77.663900      122714   
14          14  Eastern              Hoodi  12.992200  77.715900      330409   
15          15  Eastern   Krishnarajapuram  13.004000  77.687800      351936   
16          16  Eastern       Mahadevapura  12.991300  77.687400      905568   
17          17  Eastern       Marathahalli  12.959200  77.697400      249182   
18          18  Eastern            Varthur  12.938900  77.741200      546186   
19          19  Eastern         Whitefield  12.969800  77.749900       83029   

       City  AverageIncome  
0   Gurgaon   18944.099790  
1   Gurgaon   56837.022200  
2   Gurgaon   41991.817440  
3   Gurgaon    6667.447632  
4   Gurgaon   53270.063890  
5   Gurgaon   50712.430220  
6   Gurgaon   60967.535870  
7   Gurgaon   59943.541560  
8   Gurgaon   58407.090340  
9   Gurgaon   55850.962100  
10  Gurgaon   41007.219540  
11  Gurgaon   26168.448090  
12  Gurgaon    7227.731930  
13  Gurgaon   54335.368710  
14  Gurgaon   22591.063480  
15  Gurgaon   36934.737730  
16  Gurgaon   35915.973330  
17  Gurgaon   58448.658520  
18  Gurgaon   36433.267300  
19  Gurgaon   44637.984600

In [60]:
import os

In [61]:
df.drop('Unnamed: 0',axis=1,inplace=True)
df.head()

Borough      Neighborhoods   Latitude  Longitude  Population     City  \
0  Central    Cantonment area  12.972442  77.580643      866377  Gurgaon   
1  Central             Domlur  12.960992  77.638726      743186  Gurgaon   
2  Central        Indiranagar  12.971891  77.641151      474289  Gurgaon   
3  Central  Jeevanbheemanagar  12.962900  77.659500      527874  Gurgaon   
4  Central        Malleswaram  13.003100  77.564300      893629  Gurgaon   

   AverageIncome  
0   18944.099790  
1   56837.022200  
2   41991.817440  
3    6667.447632  
4   53270.063890

In [62]:
url ="https://raw.githubusercontent.com/DeepentiA/Capstone_Project/master/Gurgaon_population.csv"
df1= pd.read_csv(url)
df1.head()

Borough      Neighborhoods  Population
0  Central    Cantonment area      866377
1  Central             Domlur      743186
2  Central        Indiranagar      474289
3  Central  Jeevanbheemanagar      527874
4  Central        Malleswaram      893629

In [63]:
url="https://raw.githubusercontent.com/DeepentiA/Capstone_Project/master/Gurgaon_income.csv"
df2 = pd.read_csv(url)
df2.head()

Borough      Neighborhoods  AverageIncome
0  Central    Cantonment area   18944.099790
1  Central             Domlur   56837.022200
2  Central        Indiranagar   41991.817440
3  Central  Jeevanbheemanagar    6667.447632
4  Central        Malleswaram   53270.063890

In [64]:
!pip install geopy
!pip install beautifulsoup4

In [65]:
import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from bs4 import BeautifulSoup # library to parse HTML and XML documents

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print("Libraries are imported.")

Libraries are imported.


In [66]:
address = 'Gurgaon'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Gurgaon are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Gurgaon are 28.4646148, 77.0299194.


In [67]:
map_GGN = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for latitude,longitude, Borough, Neighborhoods in zip(df['Latitude'],df['Longitude'], df['Borough'], df['Neighborhoods']):
    label = '{}, {}'.format(Neighborhoods, Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
     parse_html=False).add_to(map_GGN)  
    
map_GGN

In [68]:
CLIENT_ID = 'AJ5GIVG4RTCVSPB1CQBINDQIL435MQJ5X32O3DX2C2SL2NZM'
CLIENT_SECRET = 'HT0QEMRESM552MVJNVZEX44HOYVPHNLYM0HZDRQJTH3MFIOQ'
VERSION = '20180606'
LIMIT = 150

In [69]:
gurgaon_latitude = df['Latitude'].mean()
gurgaon_longitude = df['Longitude'].mean()
print("Latitude and Longitude of Gurgaon are : ",gurgaon_latitude,gurgaon_longitude)

Latitude and Longitude of Gurgaon are :  12.9623396203125 77.601752946875


In [70]:
unique_boroughs_of_gurgaon = df['Borough'].unique().tolist


In [71]:
unique_boroughs_of_gurgaon

<function ndarray.tolist>

In [72]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import requests
import folium 
import seaborn as sns
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans

In [73]:
borough_colors ={}
for i in unique_boroughs_of_gurgaon:
    borough_colors[i] = '#%02X%02X%02X' % tuple(np.random.choice(range(256),size=3))

TypeError: 'builtin_function_or_method' object is not iterable

{'Central': '#2ECA6E',
 'Eastern': '#0E1B15',
 'NorthEastern': '#B12985',
 'Northern': '#AF4FF5',
 'SouthEastern': '#68B1E1',
 'Southern': '#EACF24',
 'SouthernSuburbs': '#F29B23',
 'Western': '#BD51B8'}

In [ ]:
gurgaon_map = folium.Map(location=[gurgaon_latitude,gurgaon_longitude],zoom_start=12,control_scale=True)

In [ ]:
for lat,lng,boro,nei in zip(df['Latitude'],
                           df['Longitude'],
                           df['Borough'],
                           df['Neighborhoods']):
    label_text = boro + ' - ' + nei
    label = folium.Popup(label_text,parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    tooltip = label_text,
    radius = 4,
    popup = label,
    color=borough_colors[boro],
    fill=True,
    fill_color = borough_colors[boro],
    fill_opacity=0.7).add_to(gurgon_map)

### **Exploring Gurgaon Neighborhoods using FourSquare API**

In [ ]:
def getNearbyVenues(names, boro, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, boro, lat, lng in zip(names, boro, latitudes, longitudes):
        print("Fetching venues for : ",name)            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&
        client_secret={}&v={}&ll={},{}&
        radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
 
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            boro,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhoods',
                             'Borough',
                             'Neighborhood Latitude', 
                             'Neighborhood Longitude', 
                             'Venue', 
                             'Venue Latitude', 
                             'Venue Longitude', 
                             'Venue Category']
    
    return(nearby_venues)

In [ ]:
gurgaon_venues = getNearbyVenues(names=df['Neighborhoods'],latitudes=df['Latitude'],longitudes=df['Longitude'],boro=df['Borough'])

Fetching venues for :  Cantonment area
Fetching venues for :  Domlur
Fetching venues for :  Indiranagar
Fetching venues for :  Jeevanbheemanagar
Fetching venues for :  Malleswaram
Fetching venues for :  Pete area
Fetching venues for :  Rajajinagar
Fetching venues for :  Sadashivanagar
Fetching venues for :  Seshadripuram
Fetching venues for :  Shivajinagar
Fetching venues for :  Ulsoor
Fetching venues for :  Vasanth Nagar
Fetching venues for :  Bellandur
Fetching venues for :  CV Raman Nagar
Fetching venues for :  Hoodi
Fetching venues for :  Krishnarajapuram
Fetching venues for :  Mahadevapura
Fetching venues for :  Marathahalli
Fetching venues for :  Varthur
Fetching venues for :  Whitefield
Fetching venues for :  Banaswadi
Fetching venues for :  HBR Layout
Fetching venues for :  Horamavu
Fetching venues for :  Kammanahalli
Fetching venues for :  Lingarajapuram
Fetching venues for :  Ramamurthy Nagar
Fetching venues for :  Hebbal
Fetching venues for :  Jalahalli
Fetching venues for :  Mathikere
Fetching venues for :  Peenya
Fetching venues for :  R. T. Nagar
Fetching venues for :  Vidyaranyapura
Fetching venues for :  Yelahanka
Fetching venues for :  Yeshwanthpur
Fetching venues for :  Bommanahalli
Fetching venues for :  Bommasandra
Fetching venues for :  BTM Layout
Fetching venues for :  Electronic City
Fetching venues for :  HSR Layout
Fetching venues for :  Koramangala
Fetching venues for :  Madiwala
Fetching venues for :  Banashankari
Fetching venues for :  Basavanagudi
Fetching venues for :  Girinagar
Fetching venues for :  J. P. Nagar
Fetching venues for :  Jayanagar
Fetching venues for :  Kumaraswamy Layout
Fetching venues for :  Padmanabhanagar
Fetching venues for :  Uttarahalli
Fetching venues for :  Anjanapura
Fetching venues for :  Arekere
Fetching venues for :  Begur
Fetching venues for :  Gottigere
Fetching venues for :  Hulimavu
Fetching venues for :  Kothnur
Fetching venues for :  Basaveshwaranagar
Fetching venues for :  Kamakshipalya
Fetching venues for :  Kengeri
Fetching venues for :  Mahalakshmi Layout
Fetching venues for :  Nagarbhavi
Fetching venues for :  Nandini Layout
Fetching venues for :  Nayandahalli
Fetching venues for :  Rajarajeshwari Nagar
Fetching venues for :  Vijayanagar


In [ ]:
gurgaon_venues.head(5)

	Neighborhood	|Borough	Neighborhood Latitude	Neighborhood Longitude	Venue	Venue Latitude	Venue Longitude	Venue Category
0	Cantonment area	Central	12.972442	77.580643	Hotel Fishland	12.975569	77.578592	Seafood Restaurant
1	Cantonment area	Central	12.972442	77.580643	Sapna Book House	12.976355	77.578461	Bookstore
2	Cantonment area	Central	12.972442	77.580643	Vasudev Adigas	12.973707	77.579257	Indian Restaurant
3	Cantonment area	Central	12.972442	77.580643	Adigas Hotel	12.973554	77.579161	Restaurant
4	Cantonment area	Central	12.972442	77.580643	Kamat Yatrinivas	12.975985	77.578125	Indian Restaurant

In [ ]:
gurgaon_venues.groupby('Venue Category').count()['Neighborhood'].sort_values(ascending=False).head(10)

Venue Category
Indian Restaurant       114
Café                     33
Bakery                   28
Fast Food Restaurant     22
Ice Cream Shop           19
Pizza Place              17
Department Store         16
Restaurant               15
Coffee Shop              15
Chinese Restaurant       13
Name: Neighborhood, dtype: int64

In [ ]:
gurgaon_venues.groupby('Venue Category').count()['Neighborhood'].sort_values(ascending=False).head(10)

Venue Category
Indian Restaurant       114
Café                     33
Bakery                   28
Fast Food Restaurant     22
Ice Cream Shop           19
Pizza Place              17
Department Store         16
Restaurant               15
Coffee Shop              15
Chinese Restaurant       13
Name: Neighborhood, dtype: int64

In [ ]:
print("Total number of unique categories in gurgaon are : ",len(bangalore_venues['Venue Category'].unique()))

Total number of unique categories in gurgaon are :  132

**Getting number of venues per neighborhood**

In [ ]:

individual_gurgaon_venue_count.rename(columns={'Borough':'NumberOfVenues'},inplace=True)

In [ ]:
sns.set(rc={'figure.figsize':(11,16)})
plot = sns.barplot(x='Neighborhood',y='NumberOfVenues',data=individual_gurgaon_venue_count)
plot.set_xticklabels(individual_gurgaon_venue_count['Neighborhood'],rotation=90)

In [79]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://github.com/DeepentiA/Capstone_Project/blob/master/download%20(1).png?raw=true")

In [ ]:
number_of_top_venues = 5

In [ ]:
for hood in bangalore_venues_grouped['Neighborhood']:
    print('---------',hood,'---------')
    temp = bangalore_venues_grouped[bangalore_venues_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['Venue','Frequency']
    temp = temp.iloc[1:]
    temp['Frequency'] = temp['Frequency'].astype(float)
    temp = temp.round({'Frequency': 2})
    print(temp.sort_values('Frequency', ascending=False).reset_index(drop=True).head(number_of_top_venues))
    print('\n')

**Frequency of each neighborhood and its top 5 venues can be known. This is very important is we can analyze top neighborhoods with most busy restaurants**

**Next challenge is to find the optimal k value for clustering and we do it using the elbow method**

In [ ]:
sse = {}
for k in range(1,15):
    kmeans = KMeans(n_clusters=k,random_state=0)
    kmeans.fit(gurgaon_venues_grouped.drop('Neighborhoods',axis=1))
    bangalore_venues_grouped['Cluster'] = kmeans.labels_
    sse[k] = kmeans.inertia_

plt.figure()
plt.plot(list(sse.keys()), list(sse.values()))
plt.xlabel("Number of cluster")
plt.ylabel("SSE")
plt.show()


In [78]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://github.com/DeepentiA/Capstone_Project/blob/master/download.png?raw=true")

## **From the above graph, we can see the optimal value for cluster is 5**

In [ ]:
kmeans = KMeans(n_clusters=5,random_state=0)

In [ ]:
kmeans.fit(gurgaon_venues_grouped.drop('Neighborhoods',axis=1))

In [ ]:
gurgaon_venues_grouped.groupby('Cluster')['Neighborhoods'].count()

In [ ]:
gurgaon_venues_grouped.columns

In [ ]:
# create map
map_clusters = folium.Map(location=[gurgaon_latitude, gurgaon_longitude], zoom_start=11)

In [ ]:
# set color scheme for the clusters
x = np.arange(6)
ys = [i + x + (i*x)**2 for i in range(6)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(neighborhoods_venues_sorted['Neighborhood Latitude'], neighborhoods_venues_sorted['Neighborhood Longitude'], neighborhoods_venues_sorted['Neighborhood'], neighborhoods_venues_sorted['Cluster']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)

In [ ]:
top3 = recommended_neighborhoods.groupby(['Neighborhoods','1st Most Common Venue','2nd Most Common Venue','3rd Most Common Venue'])['Ranking'].unique()

In [ ]:
top3_df = pd.DataFrame(top3).reset_index()

In [86]:
url= "https://raw.githubusercontent.com/DeepentiA/Capstone_Project/master/best%20food.csv"
df5 = pd.read_csv(url)
df5.head()

Neighborhoods                   1st Most Common Venue  \
0          Begur                   Venue Category_Bakery   
1       Hulimavu                   Venue Category_Bakery   
2  Kamakshipalya  Venue Category_South Indian Restaurant   

              2nd Most Common Venue           3rd Most Common Venue  \
0  Venue Category_Indian Sweet Shop          Venue Category_Stadium   
1  Venue Category_Indian Restaurant  Venue Category_Badminton Court   
2        Venue Category_Yoga Studio       Venue Category_Food Truck   

                Ranking  
0  [0.7361321887351776]  
1  [0.7638135476902764]  
2    [0.80418735993893]

### Conclusion :
**Here, according the data, we see that 'Begur,Hulimavu and kamakshipalya' are top 3 neighborhoods to find similar food to whitefield restaurant's veg food !
Here our model will recommend these neighborneeds and top 3 common venues to visit**